### PR TITLE
Move JLanguage::exists to JLanguageHelper::exists

### PR DESCRIPTION
**Pull request by andrepereiradasilva**
### Summary of Changes

This PR moves `JLanguage::exists` to `JLanguageHelper::exists` since this method does not needs to be in JLanguage as this method doesn't do any interactions with JLanguage class neither depends on it.
Also deprecates `JLanguage::exists` and do all necessary replacements in the core.
### Testing Instructions
1. Apply patch 
2. Test if Jooma works fine
3. Code review.
### Documentation Changes Required

None that i know of.

### CODE REVIEW

File: libraries/joomla/language/helper.php [hunk 1]
```diff
@@ -71,6 +71,42 @@ public static function createLanguageList($actualLanguage, $basePath = JPATH_BAS
...
+	 * Checks if a language exists.
+	 *
+	 * This is a simple, quick check for the directory that should contain language files for the given user.
+	 *
+	 * @param   string  $lang      Language to check.
+	 * @param   string  $basePath  Optional path to check.
+	 *
+	 * @return  boolean  True if the language exists.
+	 *
+	 * @since   11.1
```
Comment (zero-24) [inline libraries/joomla/language/helper.php]: This should be deploy version i guess ;)

Comment (andrepereiradasilva) [inline libraries/joomla/language/helper.php]: done

File: libraries/joomla/language/language.php [hunk 2]
```diff
@@ -683,29 +683,12 @@ public function setSearchDisplayedCharactersNumberCallback($function)
...
-		if (isset($paths[$path]))
-		{
-			return $paths[$path];
-		}
-
-		// Check if the language exists
-		$paths[$path] = is_dir($path);
-
-		return $paths[$path];
+		return JLanguageHelper::exists($lang, $basePath);
```
Comment (C-Lodder) [inline libraries/joomla/language/language.php]: Would it also be good to add `JLog::add('The use of JLanguage::exists() is deprecated use JLanguageHelper::exists instead.', JLog::WARNING, 'deprecated');`

Comment (andrepereiradasilva) [inline libraries/joomla/language/language.php]: done

**Comment by infograf768** [comment]
Looks OK here.

**Comment by andrepereiradasilva** [comment]
BTW this also saves 1.2 ms (of a total page generate of 137.15 ms) during `JApplicationSite::initializeApp`, ie, all site pages (in my php7 multilingual site).

### Before
image

### After
image

If you want to make this performance test, enable debug system plugin and in https://github.com/joomla/joomla-cms/blob/staging/libraries/cms/application/site.php#L582 add
```php
	protected function initialiseApp($options = array())
	{
		// [Code removed for brevity]

		if (JPluginHelper::isEnabled('system', 'languagefilter'))
		{
			// [Code removed for brevity]
		}

		!JDEBUG ?: $this->profiler->mark('-> initialiseSiteApp (languages part) START');

		// [Code removed for brevity]

		!JDEBUG ?: $this->profiler->mark('-> initialiseSiteApp (languages part) END');

		// Finish initialisation
		parent::initialiseApp($options);
	}
```

**Comment by andrepereiradasilva** [comment]
then try it yourself

**Comment by andrepereiradasilva** [comment]
> 1.2ms is negligible and most probably

"negligible", well it depends. when you're all site loads in 130ms (like mine does), it's almosts 1% of the total ... so do 100 performance improvements like this one and you get a total server page generation time of 1ms :wink: .

**Comment by compojoom** [comment]
Okay, I take it back. It seems that it really improves performance.


image

Any ideas why this is actually happening? What's the hidden PHP optimization that we are hitting with that?

**Comment by andrepereiradasilva** [comment]
First, let me say i'm not a full time PHP developer (it's more like an hobby in my case) so i could be wrong here.

So, taking this case as an example, i guess that since we are now directly calling the JLanguageHelper class, php doesn't need to instanciate the whole JLanguage class to perform this particular operation that in reality does not depend on JLanguage class methods or properties. So, probably, that's why we have a slight performance/memory usage improvement.

Since i noticed the joomla framework is in the process of doing the same for this and other JLanguage methods (https://github.com/joomla-framework/language/blob/2.0-dev/src/LanguageHelper.php#L30), for sure @mbabker can explain better than i do why this slight performance/memory usage improvement happens.

**Comment by compojoom** [comment]
I thought the same, but isn't the JLanguage class going to be called at a later stage, because joomla needs it for the language anyway. So the seconds gained at this stage had to be lost later when the jlanguage class is initialised. As you can see from the total before and after - this doesn't seem to be the case. So, I doubt that this is what's going on here.

**Comment by andrepereiradasilva** [comment]
I think the current selected JLanguage instance is created exactly after this code in the parent `JApplicationCms::initialiseApp` method.

See https://github.com/joomla/joomla-cms/blob/staging/libraries/cms/application/cms.php#L638

**Comment by andrepereiradasilva** [comment]
i did some test and it seems you are right there is no significant change since the performance and memory usage is transfered to `JApplicationCms::initialiseApp` `JLanguage::getInstance` method.

#### Before patch
image

#### After patch
image

**Comment by mbabker** [comment]
Really the only performance benefit here is when using the method without having a `JLanguage` instance already instantiated into memory.  Which from an architecture standpoint is pretty good, but when it comes to the overall performance of the application and its handling of the request, nobody's going to notice the difference because so many of the global services are instantiated and in use before you even get to the first plugin event (by the time `onAfterInitialise` triggers, you've already got `JSession` running with the session started, your database connection open because of the app's `checkSession()` method or having read your session data into memory on the default configuration, an open connection to the cache storage adapter, the global `JLanguage` instance with correct language configured, among others; [this gist](https://gist.github.com/mbabker/04dc4b6f2d1774352a99b9a4e45d9f75) is a snapshot of everything that's loaded by that point from a few months ago).

It's one of those good micro optimizations, but the way the CMS is structured right now it's one that is negligible at best.

**Comment by andrepereiradasilva** [comment]
ok, i understood. From an architecture standpoint it makes sense but will not have a performance benefit because JLanguage is always loaded in the app.

Anyway, do you have plans for the CMS to start using Language and LanguageHelper from the framework just like we use, for instance, ArrayHelper now?

**Comment by mbabker** [comment]
I have no idea what to do about that.  The Framework's Language package has changes that make using it in the CMS less favorable, like not supporting loading files from multiple applications (the JLanguage constructor has path lookups for `JPATH_SITE` and `JPATH_ADMINISTRATOR` while the Framework only supports `JPATH_ROOT` in v1 and requires a base path to be injected in v2).  Until the CMS has proper service driven architecture there's probably a slim chance if any to ever use the Framework package.

**Comment by andrepereiradasilva** [comment]
ok, i see, so 5.0 maybe ;)
thanks again for your help and explanations michael.

**Comment by zero-24** [comment]
I have tested this item :white_check_mark: successfully on 7fca9338f819cb4288cd81854bcc6eec3cbe3db7<br /><br />Works for me navigating to the frontend and backend as well as in different langauges. Thanks.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11986">issues.joomla.org/joomla-cms/11986</a>.</sub>

**Pull request closed without merging**